File: lecture04/02_happy/happy_number.py
#!/usr/bin/env python3

import sys

# Naive version

def is_happy(n, seen):
    if n in seen:                   # Base case: cycles
        return False

    if n == 1:                      # Base case: reaches 1
        return True

    seen.add(n)                     # Mark seen

    squares = sum(x * x for x in map(int, str(n)))
    return is_happy(squares, seen)  # Recurse

# Memoized version

IsHappy = {}                        # Memoization cache

def is_happy_cached(n, seen):
    if n in seen:                   # Base case: cycles
        return False

    if n == 1:                      # Base case: reaches 1
        return True
    
    seen.add(n)                     # Mark seen

    if n not in IsHappy:            # Memoize recursive call
        squares    = sum(x * x for x in map(int, str(n)))
        IsHappy[n] = is_happy_cached(squares, seen)

    return IsHappy[n]

# Main execution

for n in map(int, sys.stdin):
    print('Yes' if is_happy_cached(n, set()) else 'No')

File: lecture04/02_happy/happy_number.cpp
// happy_number.cpp

#include <iostream>
#include <numeric>
#include <unordered_map>
#include <unordered_set>

using namespace std;

// Global variables

unordered_map<int, bool> IsHappy;

// Functions

bool is_happy_r(int n, unordered_set<int> &seen) {
    if (seen.count(n)) {	// Base case: cycles
    	return false;
    }
    
    if (n == 1) {		// Base case: reaches 1
    	return true;
    }

    seen.insert(n);		// Mark seen

    if (!IsHappy.count(n)) {	// Memoize recursive call
    	string s = to_string(n);
    	int squares = accumulate(s.begin(), s.end(), 0, [](int sum, char c) {
    	    int d = (c - '0');
    	    return sum + (d*d);
	});
	IsHappy[n]  = is_happy_r(squares, seen);
    }

    return IsHappy[n];
}

bool is_happy(int n) {
    unordered_set<int> seen;
    return is_happy_r(n, seen);
}

// Main execution

int main(int argc, char *argv[]) {
    int n;

    while (cin >> n) {
    	cout << (is_happy(n) ? "Yes" : "No") << endl;
    }

    return 0;
}
